File: 0341-Nested-Iterator/0341.js
// 341. 扁平化嵌套列表迭代器
// 给你一个嵌套的整型列表。请你设计一个迭代器，使其能够遍历这个整型列表中的所有整数。

// 列表中的每一项或者为一个整数，或者是另一个列表。其中列表的元素也可能是整数或是其他列表。



// 示例 1:

// 输入: [[1, 1], 2, [1, 1]]
// 输出: [1, 1, 2, 1, 1]
// 解释: 通过重复调用 next 直到 hasNext 返回 false，next 返回的元素的顺序应该是: [1, 1, 2, 1, 1]。
// 示例 2:

// 输入: [1, [4, [6]]]
// 输出: [1, 4, 6]
// 解释: 通过重复调用 next 直到 hasNext 返回 false，next 返回的元素的顺序应该是: [1, 4, 6]。

/**
 * @constructor
 * @param {NestedInteger[]} nestedList
 */
var NestedIterator = function (nestedList) {
  this.stack = nestedList;
};


/**
 * @this NestedIterator
 * @returns {boolean}
 */
NestedIterator.prototype.hasNext = function () {
  while (this.stack.length !== 0) {
    if (this.stack[0].isInteger()) {
      return true;
    } else {
      let cur = this.stack[0].getList();
      this.stack.shift();
      this.stack.unshift(...cur);
    }
  }
};

/**
 * @this NestedIterator
 * @returns {integer}
 */
NestedIterator.prototype.next = function () {
  return this.stack.shift().getInteger();
};

File: 0341-Nested-Iterator/0341.java
public class NestedIterator implements Iterator<Integer> {
  // 存储列表当前的遍历位置
  private Deque<Iterator<NestedIterator>> stack;

  public NestedIterator(List<NestedInteger> nestedList) {
    stack = new LinkedList<Iterator<NestedInteger>>();
    stack.push(nestedList.iterator());
  }

  @override
  public Integer next() {
    // 由于保证调用next之前会调用hasNext，直接返回栈顶列表的当前元素
    return stack.peek().next().getInteger();
  }

  @override
  public boolean hasNext() {
    while (!stack.isEmpty()) {
      Iterator<NestedInteger> it = stack.peek();
      if (!it.hasNext()) {
        // 遍历到当前列表末尾， 出栈
        stack.pop();
        continue;
      }
      // 若取出的元素是整数，则通过创建一个额外的列表将其重新放入栈中
      NestedInteger nest = it.next();
      if (nest.isInteget()) {
        List<NestedInteger> list = new ArrayLists<NestedInteger>();
        list.add(nest);
        stack.push(list.iterator());
        return true;
      }
      stack.push(nest.getList().iterator());
    }
    return false;
  }
}